### DISCUSSION TITLE
Best way to develop the Phoniebox code

### DISCUSSION THREAD

**Discussion by pabera** [MiczFlor/RPi-Jukebox-RFID#1305] [Q&A]
I wanted to know from those who have developed or contributed to this project, how you guys have set up your development environment?

Are you guys always ssh'ing into your Raspberry Pi and clone the Git repo there? Or do you develop on your local environment with some sort of sync or SSHFS? Is Docker an option?

Patrick

**Comment by s-martin** [comment]
I use Visual Studio Code and the extensions Python and Remote SSH to develop directly on the Pi.

But currently my dev Pi is broken :-(

**Comment by arne123** [comment]
This is a very good question, I think it would be even worth a best practice chapter somewhere.
I do like the visual studio approach, unfortunately it is not working on every target.

I do sshfs + ssh on the target, but I would like to work offline also since the Box it is not always available 
(well it had some reason why I build it)

I have seen that there are a lot of articles about stubbing gpio or fake-gpio especially for python.
So to to create a kind of sandbox to test and develop the Phoniebox code.
Does somebody already have some experience with this approach?

**Reply by s-martin** [reply]
I searched around a little to simulate or fake a Phoniebox, but never really finished.

But I would love such a „Phoniebox Simulator“, I would be willing to help and develop such a simulator.

Maybe we could use docker or it would be perfect to work with Windows WSL.

#711 and #715 also try to address some improvements.

**Reply by pabera** [reply]
Thanks you two for this insights and references. With a little research I think this should be doable. There are a few resources on the web explaining how to set up RPi OS on Virtual Box. The only missing pieces now are GPIO and RFID. At least for GPIO, this lib could help: [fake-rpigpio](https://pypi.org/project/fake-rpigpio/). Am I missing something?

**Reply by arne123** [reply]
For Rfid I would propose to configure a dummy reader class which react on keyboard inputs (terminal or even tkinter etc.)

**Reply by s-martin** [reply]
So I tried the folllowing on Windows 10:
- Installing Ubuntu 20.04 for WSL
- Using Visual Studio Code with Remote WSL and Python extension
- Installing python3-pip
- Installing fake-rpigpio and evdev from pypi

Using 
```
try:
    import RPi.GPIO
except (RuntimeError, ModuleNotFoundError):
    import fake_rpigpio.utils
    fake_rpigpio.utils.install()
```
in `Reader.py.experimental` worked, but you can't do anything really

If you try to run (or debug) `daemon_rfid_reader.py`, you need to run `RegisterDevice.py` first to configure a reader. So it would make sense to have a dummy reader.

**Reply by s-martin** [reply]
- Create a file scripts/deviceName.txt with content "Dummy"
- make changes in https://github.com/s-martin/RPi-Jukebox-RFID/tree/createTestEnv to Reader.py.experimental (and you need to rename to Reader.py)
- Debug daemon_rfid_reader.py 

But so far there's not much meaningful to do.

**Reply by MiczFlor** [reply]
I spent a lot of time trying to get a Virtual Box (and there are other virtualisations) version of the Pi OS running. It would make life so much easier :) But I failed, so far. Having said this, the last try was possibly a year ago.
